app/src/main/java/org/thoughtcrime/securesms/contacts/ContactAccessor.java

```diff
@@ -0,0 +1,340 @@
+      dest.writeString(number);
+      dest.writeString(type);
+    }
+  }
+
+  public static class ContactData implements Parcelable {
+
+    public static final Parcelable.Creator<ContactData> CREATOR = new Parcelable.Creator<ContactData>() {
+      public ContactData createFromParcel(Parcel in) {
+        return new ContactData(in);
+      }
+
+      public ContactData[] newArray(int size) {
+        return new ContactData[size];
+      }
+    };
+
+    public final long id;
+    public final String name;
+    public final List<NumberData> numbers;
+
+    public ContactData(long id, String name) {
+      this.id      = id;
+      this.name    = name;
+      this.numbers = new LinkedList<NumberData>();
+    }
+
+    public ContactData(Parcel in) {
+      id      = in.readLong();
+      name    = in.readString();
+      numbers = new LinkedList<NumberData>();
+      in.readTypedList(numbers, NumberData.CREATOR);
+    }
+
+    public int describeContents() {
+      return 0;
+    }
+
+    public void writeToParcel(Parcel dest, int flags) {
+      dest.writeLong(id);
+      dest.writeString(name);
+      dest.writeTypedList(numbers);
+    }
+  }
+
+  /***
+   * If the code below looks shitty to you, that's because it was taken
+   * directly from the Android source, where shitty code is all you get.
+   */
+
+  public Cursor getCursorForRecipientFilter(CharSequence constraint,
+      ContentResolver mContentResolver)
+  {
+    final String SORT_ORDER = Contacts.TIMES_CONTACTED + " DESC," +
+                              Contacts.DISPLAY_NAME + "," +
+                              Contacts.Data.IS_SUPER_PRIMARY + " DESC," +
+                              Phone.TYPE;
+
+    final String[] PROJECTION_PHONE = {
+        Phone._ID,                  // 0
+        Phone.CONTACT_ID,           // 1
+        Phone.TYPE,                 // 2
+        Phone.NUMBER,               // 3
+        Phone.LABEL,                // 4
+        Phone.DISPLAY_NAME,         // 5
+    };
+
+    String phone = "";
+    String cons  = null;
+
+    if (constraint != null) {
+      cons = constraint.toString();
+
+      if (RecipientsAdapter.usefulAsDigits(cons)) {
+        phone = PhoneNumberUtils.convertKeypadLettersToDigits(cons);
+        if (phone.equals(cons) && !PhoneNumberUtils.isWellFormedSmsAddress(phone)) {
+          phone = "";
+        } else {
+          phone = phone.trim();
+        }
+      }
+    }
+    Uri uri = Uri.withAppendedPath(Phone.CONTENT_FILTER_URI, Uri.encode(cons));
+    String selection = String.format("%s=%s OR %s=%s OR %s=%s",
+                                     Phone.TYPE,
+                                     Phone.TYPE_MOBILE,
+                                     Phone.TYPE,
+                                     Phone.TYPE_WORK_MOBILE,
+                                     Phone.TYPE,
+                                     Phone.TYPE_MMS);
+
+    Cursor phoneCursor = mContentResolver.query(uri,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *phoneCursor*. The resource is referenced at line 334. The resource is returned at line 336. There are other execution paths that don&#39;t close the resource or return it, for example, when an if condition at line 314 evaluates to true. Ensure *phoneCursor* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).